### PR TITLE
Fogl 3648:: handling race conditions that produce the message SQLITE_BUSY 

### PR DESCRIPTION
**Pull request by sstefano69**

### CODE REVIEW

File: C/plugins/storage/sqlite/common/readings.cpp [hunk 1]
```diff
@@ -495,19 +497,22 @@ int Connection::readingStream(ReadingStream **readings, bool commit)
 
 						if (sqlite3_resut == SQLITE_LOCKED  )
 						{
-							sleep_time_ms = (1 * PREP_CMD_RETRY_BACKOFF);
+							sleep_time_ms = PREP_CMD_RETRY_BASE + (random() &  PREP_CMD_RETRY_BACKOFF);
```
Comment (MarkRiddoch) [inline C/plugins/storage/sqlite/common/readings.cpp]: Do you mean to do a logical add here? You are doing an AND operation with 0x1388 which only gives you small range of values. Is it perhaps (random() % PREP_CMD_RETRY_BACKOFF) you want to give you a value between 0 and 4999?

File: C/plugins/storage/sqlite/common/readings.cpp [hunk 2]
```diff
@@ -495,19 +497,22 @@ int Connection::readingStream(ReadingStream **readings, bool commit)
...
 
 							std::this_thread::sleep_for(std::chrono::milliseconds(sleep_time_ms));
 						}
 						if (sqlite3_resut == SQLITE_BUSY)
 						{
-							sleep_time_ms = (1 * PREP_CMD_RETRY_BACKOFF);
+							ostringstream threadId;
+							threadId << std::this_thread::get_id();
+
+							sleep_time_ms = PREP_CMD_RETRY_BASE + (random() &  PREP_CMD_RETRY_BACKOFF);
```
Comment (MarkRiddoch) [inline C/plugins/storage/sqlite/common/readings.cpp]: Sane question as above.

File: C/plugins/storage/sqlite/common/readings.cpp [hunk 3]
```diff
@@ -719,19 +724,22 @@ int sleep_time_ms = 0;
 					}
 					if (sqlite3_resut == SQLITE_LOCKED  )
 					{
-						sleep_time_ms = (1 * PREP_CMD_RETRY_BACKOFF);
+						sleep_time_ms = PREP_CMD_RETRY_BASE + (random() &  PREP_CMD_RETRY_BACKOFF);
```
Comment (MarkRiddoch) [inline C/plugins/storage/sqlite/common/readings.cpp]: And same again

File: C/plugins/storage/sqlite/common/readings.cpp [hunk 4]
```diff
@@ -719,19 +724,22 @@ int sleep_time_ms = 0;
...
 
 						std::this_thread::sleep_for(std::chrono::milliseconds(sleep_time_ms));
 					}
 					if (sqlite3_resut == SQLITE_BUSY)
 					{
-						sleep_time_ms = (1 * PREP_CMD_RETRY_BACKOFF);
+						ostringstream threadId;
+						threadId << std::this_thread::get_id();
+
+						sleep_time_ms = PREP_CMD_RETRY_BASE + (random() &  PREP_CMD_RETRY_BACKOFF);
```
Comment (MarkRiddoch) [inline C/plugins/storage/sqlite/common/readings.cpp]: Same issue with & versus %

**Review by MarkRiddoch — CHANGES_REQUESTED** [review]
Probably a good idea to call srandom somewhere, maybe use getpid() or time as the seed value.

File: C/plugins/storage/sqlite/common/readings.cpp (hunk 1 above)
Comment (sstefano69) [reply C/plugins/storage/sqlite/common/readings.cpp]: the intention is to have 
- a fixed amount of time to wait PREP_CMD_RETRY_BASE 
- plus a randomly generated: (random() &  PREP_CMD_RETRY_BACKOFF) - produces a number between 0 to 5000

some logs:

```
... sleep time ms :5896:
... sleep time ms :9104:
... sleep time ms :5000:
... sleep time ms :9096:

```

Comment (sstefano69) [reply C/plugins/storage/sqlite/common/readings.cpp]: fixed

File: C/plugins/storage/sqlite/common/readings.cpp (hunk 2 above)
Comment (sstefano69) [reply C/plugins/storage/sqlite/common/readings.cpp]: fixed

File: C/plugins/storage/sqlite/common/readings.cpp (hunk 3 above)
Comment (sstefano69) [reply C/plugins/storage/sqlite/common/readings.cpp]: fixed

File: C/plugins/storage/sqlite/common/readings.cpp (hunk 4 above)
Comment (sstefano69) [reply C/plugins/storage/sqlite/common/readings.cpp]: fixed

**Review by MarkRiddoch — APPROVED** [review]

**Review by YashTatkondawar — CHANGES_REQUESTED** [review]
I have tested the following changes on RPi4 buster with 2GB RAM with lab script, and I can still see the errors and warning logs. So it does not solve the problem.

**Comment by sstefano69** [comment]
@YashTatkondawar 
@MarkRiddoch 

Hi Yash, I think (already talked with Mark) we can approve and close this PR as I have done more tests and after a while (1 hour and half) I saw the error but it is a different situation respect this one it is a 'database is locked' that has impacted other tables for example the 'foglamp.tasks'

btw: I tested on 2 RPs:
- one having a degraded sd card where the problem happened 
- and one having a good sd card that worked without problems at all

I'm going to create a specific  jira about that.

```
Feb  4 05:40:31 raspberrypi FogLAMP Storage[728]: ERROR: SQLite3 storage plugin raising error: database is locked
Feb  4 05:40:31 raspberrypi FogLAMP Storage[728]: ERROR: SQL statement: BEGIN TRANSACTION;UPDATE foglamp.tasks SET exit_code = 1, state = 2, end_time = '2020-02-04 05:37:07.246885+00:00' WHERE id = '692df3fd-7b7b-4e1b-a1d4-b3132f7aaa31';COMMIT TRANSACTION;
Feb  4 05:40:31 raspberrypi FogLAMP[714] ERROR: storage_client: foglamp.common.storage_client.storage_client: Error code: 400, reason: Bad Request, details: {'entryPoint': 'update', 'message': 'database is locked', 'retryable': False}
Feb  4 05:40:31 raspberrypi FogLAMP[714] ERROR: scheduler: foglamp.services.core.scheduler.scheduler: Update failed: {"values": {"exit_code": 1, "state": 2, "end_time": "2020-02-04 05:37:07.246885+00:00"}, "where": {"column": "id", "condition": "=", "value": "692df3fd-7b7b-4e1b-a1d4-b3132f7aaa31"}}#012Traceback (most recent call last):#012  File "/usr/local/foglamp/python/foglamp/services/core/scheduler/scheduler.py", line 273, in _wait_for_task_completion#012    res = await self._storage_async.update_tbl("tasks", update_payload)#012  File "/usr/local/foglamp/python/foglamp/common/storage_client/storage_client.py", line 204, in update_tbl#012    raise StorageServerError(code=resp.status, reason=resp.reason, error=jdoc)#012foglamp.common.storage_client.exceptions.StorageServerError: code: 400, reason:Bad Request, error:{'entryPoint': 'update', 'message': 'database is locked', 'retryable': False} 
```

**Review by YashTatkondawar — APPROVED** [review]
Approved this on suggestion of @sstefano69  @MarkRiddoch .

**Pull request merged**